sdk/servicebus/Microsoft.Azure.ServiceBus/src/Primitives/ConcurrentExpiringSet.cs

```diff
@@ -4,64 +4,90 @@
 
     sealed class ConcurrentExpiringSet<TKey>
     {
         readonly ConcurrentDictionary<TKey, DateTime> dictionary;
-        readonly object cleanupSynObject = new object();
-        bool cleanupScheduled;
-        static TimeSpan delayBetweenCleanups = TimeSpan.FromSeconds(30);
+        readonly ICollection<KeyValuePair<TKey, DateTime>> dictionaryAsCollection;
+        readonly CancellationTokenSource tokenSource = new CancellationTokenSource();
+        volatile TaskCompletionSource<bool> cleanupTaskCompletionSource = new TaskCompletionSource<bool>(TaskCreationOptions.RunContinuationsAsynchronously);
+        volatile int closeSignaled;
+        bool closed;
+        static readonly TimeSpan delayBetweenCleanups = TimeSpan.FromSeconds(30);
 
         public ConcurrentExpiringSet()
         {
             this.dictionary = new ConcurrentDictionary<TKey, DateTime>();
+            this.dictionaryAsCollection = dictionary;
+            _ = CollectExpiredEntriesAsync(tokenSource.Token);
         }
 
         public void AddOrUpdate(TKey key, DateTime expiration)
         {
+            this.ThrowIfClosed();
+
             this.dictionary[key] = expiration;
-            this.ScheduleCleanup();
+            this.cleanupTaskCompletionSource.TrySetResult(true);
         }
 
         public bool Contains(TKey key)
         {
+            this.ThrowIfClosed();
+
             return this.dictionary.TryGetValue(key, out var expiration) && expiration > DateTime.UtcNow;
         }
 
-        void ScheduleCleanup()
+        public void Close()
         {
-            lock (this.cleanupSynObject)
+            if (Interlocked.Exchange(ref this.closeSignaled, 1) != 0)
             {
-                if (this.cleanupScheduled || this.dictionary.Count <= 0)
-                {
-                    return;
-                }
-
-                this.cleanupScheduled = true;
-                Task.Run(async () => await this.CollectExpiredEntriesAsync().ConfigureAwait(false));
+                return;
             }
+
+            this.closed = true;
+
+            this.tokenSource.Cancel();
+            this.cleanupTaskCompletionSource.TrySetCanceled();
+            this.dictionary.Clear();
+            this.tokenSource.Dispose();
         }
 
-        async Task CollectExpiredEntriesAsync()
+        async Task CollectExpiredEntriesAsync(CancellationToken token)
         {
-            await Task.Delay(delayBetweenCleanups);
-
-            lock (this.cleanupSynObject)
+            while (!token.IsCancellationRequested)
             {
-                this.cleanupScheduled = false;
-            }
+                try
+                {
+                    await cleanupTaskCompletionSource.Task.ConfigureAwait(false);
+                    this.cleanupTaskCompletionSource = new TaskCompletionSource<bool>(TaskCreationOptions.RunContinuationsAsynchronously);
+                    await Task.Delay(delayBetweenCleanups, token).ConfigureAwait(false);
```

Comment:
Can this also throw `ObjectDisposedException`?

Reply:
I don't think it can given that dispose of the CTS only releases the timer and the linked registrations if any

Reply:
I don't completely know the inner details of CTS. I am going to trust your word that this would not throw `ObjectDisposedException` and also that   `this.dictionaryAsCollection.Remove(kvp);` will not throw while iteration/dictionary.Clear() is in progress.